File: packages/gavio-py/gavio/response.py
"""GavioResponse — the canonical response returned to the caller."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from .types import CacheType, TokenUsage

if TYPE_CHECKING:
    from .interceptors.audit.record import AuditRecord


@dataclass
class GavioResponse:
    """Result of a gateway call, enriched by the post-interceptor pipeline."""

    trace_id: str
    content: str
    model: str
    provider: str
    model_version: str = ""
    usage: TokenUsage = field(default_factory=TokenUsage)
    cost_usd: float = 0.0
    latency_ms: int = 0
    cache_hit: bool = False
    cache_type: CacheType | None = None
    interceptors_fired: list[str] = field(default_factory=list)
    audit: AuditRecord | None = None
    metadata: dict[str, Any] = field(default_factory=dict)

    def copy_with_content(self, content: str) -> GavioResponse:
        """Return a copy with replaced content (used by PII restore, guardrails)."""
        clone = GavioResponse(
            trace_id=self.trace_id,
            content=content,
            model=self.model,
            provider=self.provider,
            model_version=self.model_version,
            usage=self.usage,
            cost_usd=self.cost_usd,
            latency_ms=self.latency_ms,
            cache_hit=self.cache_hit,
            cache_type=self.cache_type,
            interceptors_fired=list(self.interceptors_fired),
            audit=self.audit,
            metadata=dict(self.metadata),
        )
        return clone

File: packages/gavio-js/src/response.ts
/** GavioResponse — the canonical response returned to the caller. */

import { TokenUsage } from './types.js'
import type { CacheType } from './types.js'
import type { AuditRecord } from './interceptors/audit/record.js'

export interface GavioResponseInit {
  traceId: string
  content: string
  model: string
  provider: string
  modelVersion?: string
  usage?: TokenUsage
  costUsd?: number
  latencyMs?: number
  cacheHit?: boolean
  cacheType?: CacheType | null
  interceptorsFired?: string[]
  audit?: AuditRecord | null
  metadata?: Record<string, unknown>
  embeddings?: number[][] | null
}

/** Result of a gateway call, enriched by the post-interceptor pipeline. */
export class GavioResponse {
  traceId: string
  content: string
  model: string
  provider: string
  modelVersion: string
  usage: TokenUsage
  costUsd: number
  latencyMs: number
  cacheHit: boolean
  cacheType: CacheType | null
  interceptorsFired: string[]
  audit: AuditRecord | null
  metadata: Record<string, unknown>
  /** Set on embedding calls (F-SEC-10): one vector per input text. */
  embeddings: number[][] | null

  constructor(init: GavioResponseInit) {
    this.traceId = init.traceId
    this.content = init.content
    this.model = init.model
    this.provider = init.provider
    this.modelVersion = init.modelVersion ?? ''
    this.usage = init.usage ?? new TokenUsage()
    this.costUsd = init.costUsd ?? 0.0
    this.latencyMs = init.latencyMs ?? 0
    this.cacheHit = init.cacheHit ?? false
    this.cacheType = init.cacheType ?? null
    this.interceptorsFired = init.interceptorsFired ?? []
    this.audit = init.audit ?? null
    this.metadata = init.metadata ?? {}
    this.embeddings = init.embeddings ?? null
  }

  /** Return a copy with replaced content (used by PII restore, guardrails). */
  copyWithContent(content: string): GavioResponse {
    return new GavioResponse({
      traceId: this.traceId,
      content,
      model: this.model,
      provider: this.provider,
      modelVersion: this.modelVersion,
      usage: this.usage,
      costUsd: this.costUsd,
      latencyMs: this.latencyMs,
      cacheHit: this.cacheHit,
      cacheType: this.cacheType,
      interceptorsFired: [...this.interceptorsFired],
      audit: this.audit,
      metadata: { ...this.metadata },
      embeddings: this.embeddings,
    })
  }
}
